src/backend/common/dispatch.hpp

```diff
@@ -42,3 +44,73 @@ inline T greatestPrimeFactor(T n) {
 
     return v;
 }
+
+// For OPENCL, the dimensions of local are returned
+// usage: cl::NDRange local = bestBlockSize<cl::NDRange>(dims, WG)
+// For CUDA, the dimensions of 1 block are returned
+// usage: dim3 block = bestBlockSize<dim3>(dims, 32);
+// The parameter dims can have any type as long as it is convertable to unsigned
+
+// Remark: The bestBlockSize is only best for independent element operations, as
+// are: copying, scaling, math on independent elements, ...
+// Since vector dimensions can be returned, it is NOT USABLE FOR BLOCK
+// OPERATIONS, as are: matmul, etc.
+template<typename Tout, typename Tin>
+Tout bestBlockSize(const Tin dims[4], unsigned warp) {
+    const unsigned d0         = static_cast<unsigned>(dims[0]);
+    const unsigned d1         = static_cast<unsigned>(dims[1]);
+    const unsigned d2         = static_cast<unsigned>(dims[2]);
+    const unsigned OCC        = 3;
```

Comment:
In the copy kernels, we only have address calculations in the threads.  This means that calculation time is very low compared to the waiting time, resulting from memory latency.

The total occupation = occupation(threads0) * occupation(threads1) * occupation(threads2).  
- Since array dim[2] is mostly very small, we will push occupation(threads2) to 100%, by only providing exact dividable block dim.
- For threads0 & threads1 occupation always drops because the last block is incomplete, although launched.  The full occupation is in brackets (assuming both equal).  The occupation per dim is calculated as follows:
 0 full blocks + 1 thread / 1 block  allocated ==> 1/32, 1/64, ...
 1 full block   + 1 thread / 2 blocks allocated ==> >50% (25%)
 2 full blocks + 1 thread / 3 blocks allocated ==> >66% (44%)
 3 full blocks + 1 thread / 4 blocks allocated ==> >75% (56%)
From blogs, I read that around 50% threads is sufficient so somewhere between 3 & 4 allocated blocks.  The performance difference between both is smaller than my measuring precision.

As result, I expect it to be mostly independent from GPU version.  On top, 
- only a very small set of dimensions are possibly affected.
- I prefer to keep it inside the function, because it is not intended to be changed outside specific performance profiling.
- There are many uncertain assumptions, so this not an exact science.  In practice, you will encounter mostly exponents of 2.
- On NVIDIA, I noticed that the driver optimizes much more so that multiple variations give the same result.


I can also remove the constant and put directly the final number in each expression.

Reply:
Updated with comments.
const unsigned OCC changed to constexpr unsigned OCCUPANCY_FACTOR